File: frontend/src/api/admin-map.ts
import { request } from "@/services/request";
import type { TencentPoiDto } from "@/api/map";

export interface AdminMapPointPhotoDto {
  photo_id: string;
  photo_type: string;
  file_url: string;
  thumbnail_url: string | null;
  caption: string | null;
  sort_order: number;
  created_at: string;
}

export interface AdminMapPointDto {
  point_id: string;
  campus_id: string;
  area_id: string | null;
  point_type: string;
  point_scope: string;
  name: string;
  subtitle: string | null;
  description: string | null;
  location_name: string | null;
  location_detail: string | null;
  lng: number;
  lat: number;
  amap_poi_id: string | null;
  amap_address: string | null;
  tencent_poi_id: string | null;
  tencent_poi_name: string | null;
  tencent_poi_address: string | null;
  tencent_poi_category: string | null;
  tencent_poi_lng: number | null;
  tencent_poi_lat: number | null;
  tencent_poi_distance_meters: number | null;
  tencent_poi_match_method: string | null;
  associated_poi: TencentPoiDto | null;
  route_instruction: string | null;
  landmark_hint: string | null;
  entrance_hint: string | null;
  icon_key: string | null;
  display_level: number;
  label_min_zoom: number;
  preview_enabled: boolean;
  preview_min_zoom: number;
  visibility: string;
  status: string;
  cover_photo_url: string | null;
  photos: AdminMapPointPhotoDto[];
  updated_at: string;
}

export interface AdminMapPointUpdatePayload {
  area_id?: string | null;
  point_type?: string;
  point_scope?: string;
  name?: string;
  subtitle?: string | null;
  description?: string | null;
  location_name?: string | null;
  location_detail?: string | null;
  amap_poi_id?: string | null;
  amap_address?: string | null;
  tencent_poi_id?: string | null;
  tencent_poi_name?: string | null;
  tencent_poi_address?: string | null;
  tencent_poi_category?: string | null;
  tencent_poi_lng?: number | null;
  tencent_poi_lat?: number | null;
  tencent_poi_distance_meters?: number | null;
  tencent_poi_match_method?: string | null;
  route_instruction?: string | null;
  landmark_hint?: string | null;
  entrance_hint?: string | null;
  icon_key?: string | null;
  display_level?: number;
  label_min_zoom?: number;
  preview_enabled?: boolean;
  preview_min_zoom?: number;
  visibility?: string;
  status?: string;
}

export interface AdminMapPointLocationPayload {
  lng: number;
  lat: number;
}

function compactPayload<T extends object>(payload: T): Record<string, unknown> {
  return Object.fromEntries(
    Object.entries(payload as Record<string, unknown>).filter(
      ([, value]) => value !== undefined,
    ),
  );
}

export function getAdminMapPoint(
  accessToken: string,
  pointId: string,
): Promise<AdminMapPointDto> {
  return request<AdminMapPointDto>({
    url: `/admin/map/points/${pointId}`,
    method: "GET",
    token: accessToken,
  });
}

export function updateAdminMapPoint(
  accessToken: string,
  pointId: string,
  payload: AdminMapPointUpdatePayload,
): Promise<AdminMapPointDto> {
  return request<AdminMapPointDto>({
    url: `/admin/map/points/${pointId}`,
    method: "PATCH",
    data: compactPayload(payload),
    token: accessToken,
  });
}

export function updateAdminMapPointLocation(
  accessToken: string,
  pointId: string,
  payload: AdminMapPointLocationPayload,
): Promise<AdminMapPointDto> {
  return request<AdminMapPointDto>({
    url: `/admin/map/points/${pointId}/location`,
    method: "PATCH",
    data: { ...payload },
    token: accessToken,
  });
}

File: backend/app/api/v1/admin_map.py
from typing import Any
from uuid import UUID

from fastapi import APIRouter, Depends, Request
from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy.orm import Session

from app.core.responses import api_success
from app.db.session import get_db
from app.modules.auth.dependencies import require_admin
from app.modules.auth.models import User
from app.modules.map import service

router = APIRouter(tags=["Admin Map"])


class AdminMapPointUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    area_id: UUID | None = None
    point_type: str | None = None
    point_scope: str | None = None
    name: str | None = Field(default=None, max_length=128)
    subtitle: str | None = Field(default=None, max_length=255)
    description: str | None = None
    location_name: str | None = Field(default=None, max_length=128)
    location_detail: str | None = None
    amap_poi_id: str | None = Field(default=None, max_length=128)
    amap_address: str | None = Field(default=None, max_length=255)
    route_instruction: str | None = None
    landmark_hint: str | None = None
    entrance_hint: str | None = None
    icon_key: str | None = Field(default=None, max_length=64)
    display_level: int | None = None
    label_min_zoom: int | None = None
    preview_enabled: bool | None = None
    preview_min_zoom: int | None = None
    visibility: str | None = None
    status: str | None = None

    def compact_payload(self) -> dict[str, Any]:
        return self.model_dump(exclude_unset=True)


class AdminMapPointLocationUpdateRequest(BaseModel):
    lng: float
    lat: float


@router.get("/points/{point_id}", summary="Get admin map point detail")
def get_admin_map_point(
    point_id: UUID,
    request: Request,
    db: Session = Depends(get_db),
    admin: User = Depends(require_admin),
):
    data = service.admin_point_detail(db, point_id=point_id)
    return api_success(data=data, trace_id=request.state.trace_id)


@router.patch("/points/{point_id}", summary="Update admin map point")
def update_admin_map_point(
    point_id: UUID,
    payload: AdminMapPointUpdateRequest,
    request: Request,
    db: Session = Depends(get_db),
    admin: User = Depends(require_admin),
):
    data = service.update_admin_point(
        db,
        point_id=point_id,
        admin=admin,
        payload=payload.compact_payload(),
    )
    return api_success(data=data, trace_id=request.state.trace_id)


@router.patch("/points/{point_id}/location", summary="Update admin map point location")
def update_admin_map_point_location(
    point_id: UUID,
    payload: AdminMapPointLocationUpdateRequest,
    request: Request,
    db: Session = Depends(get_db),
    admin: User = Depends(require_admin),
):
    data = service.update_admin_point_location(
        db,
        point_id=point_id,
        admin=admin,
        lng=payload.lng,
        lat=payload.lat,
    )
    return api_success(data=data, trace_id=request.state.trace_id)

